File: js/Tree.js


class Node {
  constructor(key) {
    this.key = key || null;
    this.p = null;
    this.rSibling = null;
    this.lChild = null;
  }
}

class BinaryNode {
  constructor(key) {
    this.key = key;
    this.p = null;
    this.left = null;
    this.right = null;
  }

  height() {
    return Math.max(
      this.left ? this.left.height() + 1 : 0,
      this.right? this.right.height() + 1 : 0
    );
  }

  size() {
    return (
      1 +
      (this.left ? this.left.size() : 0) +
      (this.right ? this.right.size() : 0)
    );
  }

  leaf() {
    return !(this.left || this.right);
  }
}

class Tree {
  constructor(root) {
    this.root = new Node();
  }
}


class BinaryTree {
  constructor(root) {
    this.root = new BinaryNode(root);
  }

  size() {
    return this.root.size();
  }

  traverseInOrder() {
    const a = [];
    function _inOrder(node) {
      if (!node) return;
      _inOrder(node.left);
      a.push(node.key);
      _inOrder(node.right);
    }
    _inOrder(this.root);
    return a;
  }

  traversePreOrder() {
    const a = [];
    function _preOrder(node) {
      if (!node) return;
      a.push(node.key);
      _preOrder(node.left);
      _preOrder(node.right);
    }
    _preOrder(this.root);
    return a;
  }

  traversePostOrder() {
    const a = [];
    function _postOrder(node) {
      if (!node) return;
      _postOrder(node.left);
      _postOrder(node.right);
      a.push(node.key);
    }
    _postOrder(this.root);
    return a;
  }
}

class BinarySearchTree extends BinaryTree {
  insert(key, node) {
    if (!this.root) {
      this.root = new BinaryNode(key);
    }

    node = node || this.root;

    if (key < node.key) {
      if (!node.left) {
        const newNode = new BinaryNode(key);
        newNode.p = node;
        node.left = newNode;
        return;
      }
      this.insert(key, node.left);
      return; // break to avoid executing below
    }

    // use above break instead of conditional to avoid another check in loop
    if (!node.right) {
      const newNode = new BinaryNode(key);
      newNode.p = node;
      node.right = newNode;
      return;
    }
    this.insert(key, node.right);
  }

  search(key, node) {
    node = node || this.root;
    if (key === node.key) {
      return node;
    } else if (key < node.key) {
      return this.search(key, node.left);
    } else {
      return this.search(key, node.right);
    }
  }

  searchIterative(key, node = this.root) {
    while (node && key !== node.key) {
      if (key < node.key) {
        node = node.left;
      } else {
        node = node.right;
      }
    }
    return node;
  }

  delete(key, node = this.root) {
    // @todo node deletion that preserves child nodes
    // need successor method
    // splice out node, find successor, etc
    // v. easy to find successor - just take minimum of right child
    // or, if no right child, walk up tree until node is no longer in right subtree
    node = this.search(key, node);

    if (!node) {
      return null;
    }

    if (!node.p) {
      this.root = null;
      return node;
    }
    
    if (key === node.p.left.key) {
      node.p.left = null;
      return node;
    }

    node.p.right = null;
    return node;
  }

  minimum(node = this.root) {
    if (!node) return null;
    if (!node.left) return node.key;
    return this.minimum(node.left);
  }

  minimumIterative(node = this.root) {
    while (node.left) {
      node = node.left;
    }
    return node.key;
  }

  maximum(node = this.root) {
    if (!node) return null;
    if (!node.right) return node.key;
    return this.maximum(node.right);
  }

  maximumIterative(node = this.root) {
    while (node.right) {
      node = node.right;
    }
    return node.key;
  }
}

class RedBlackNode extends BinaryNode {
  constructor(key) {
    super(key);
    this.color = null;
  }
}

class RedBlackTree extends BinarySearchTree {
  // @todo: when to know how to rotate?
  // @todo: when to assign color? is it purely theoretical?
  constructor(root) {
    super(root);
    this.root = new RedBlackNode(root);
    this.root.color = 'black';
  }

  lRotate(node) {
    const swap = node.right;

    // reassign subtree
    node.right = swap.left;
    if (swap.left) {
      swap.left.p = node;
    }

    // reassign parent pointers
    swap.p = node.p;
    if (!node.p) {
      this.root = swap;
    } else {
      if (Object.is(node.p.left, node)) {
        node.p.left = swap;
      } else {
        node.p.right = swap
      }
    }

    // reassign nodes' pointers
    swap.left = node;
    node.p = swap;
  }

  rRotate(node) {
    const swap = node.left;

    // reassign subtree
    node.left = swap.right;
    if (swap.right) {
      swap.right.p = node;
    }

    // reassign parent pointers
    swap.p = node.p;
    if (!node.p) {
      this.root = swap;
    } else {
      if (Object.is(node.p.left, node)) {
        node.p.left = swap;
      } else {
        node.p.right = swap
      }
    }

    // reassign nodes' pointers
    swap.right = node;
    node.p = swap;
  }

  insertFixup(node) {
    // @todo - comprehend three cases of insert-fixup
    // how have we violated red-black properties?
    // did we make the make the root red? only if inserting root.
    // did we insert a red node on a red child? probably.
    // loop over our chain of red nodes, flipping their parents black, rotating left or right as necessary while they've still got red parents.
    let parent;
    let grandparent;
    let uncle;
    console.log(`fixing up node ${node.key}`)

    while (node.p && node.p.color == 'red') {
      console.log(`while:`)
      console.log(node);
      parent = node.p;
      if (parent && !Object.is(parent, this.root)) {
        grandparent = parent.p;
        uncle = Object.is(parent, grandparent.left) ?
          grandparent.right :
          grandparent.left;
        if (uncle && uncle.color == 'red') {
          // case 1
          parent.color = 'black';
          uncle.color = 'black';
          grandparent.color = 'red';
          node = grandparent;
        } else if (parent && uncle) {
          if (Object.is(node, parent.right)) {
            node = parent;
            this.lRotate(node);
          }
          parent.color = 'black';
          grandparent.color = 'red';
          this.rRotate(grandparent)
        }
      }
    }

    // while loop with pointers not particularly well suited to JS.
    // possible to do this with a few conditions.
  }

  insert(key, node = this.root) {
    console.log(`inserting key ${key}`)
    if (!node.key) {
      node.key = key;
      return node;
    }

    if (key > node.key) {
      if (!node.right) {
        const newNode = new RedBlackNode(key);
        newNode.color = 'red';
        newNode.p = node;
        node.right = newNode;

        this.insertFixup(newNode);
        return newNode;
      }
      return this.insert(key, node.right);
    }

    if (!node.left) {
      const newNode = new RedBlackNode(key);
      newNode.color = 'red';
      newNode.p = node;
      node.left = newNode;

      this.insertFixup(newNode);
      return newNode;
    }
    return this.insert(key, node.left);
  }

  insertIterative(key) {
    const newNode = new RedBlackNode(key);
    newNode.color = 'red';

    let node = this.root;
    let prev = node;

    while (node) {
      prev = node;
      if (key > node.key) {
        node = node.right;
      } else {
        node = node.left;
      }
    }
    if (key > prev.key) {
      prev.right = newNode;
    } else {
      prev.left = newNode;
    }
    newNode.p = prev;

    this.insertFixup(newNode);
    return newNode;
  }
}

export { Node, BinaryNode, Tree, BinaryTree, BinarySearchTree, RedBlackTree };

File: python/tree.py
import unittest
# import random


class Node:
    """Base tree node"""

    def __init__(self, value):
        self.value = value
        self.p = None
        self.left = None
        self.right = None
        self.color = None

    def __str__(self):
        return '<%d>' % self.value

    def height(self):
        return max(
            self.left.height() + 1 if self.left else 1,
            self.right.height() + 1 if self.right else 1,
        )

    def minimum(self):
        if not self.left:
            return self.value
        return self.left.minimum()

    def maximum(self):
        if not self.right:
            return self.value
        return self.right.maximum()

    def successor(self):
        if self.right:
            return self.right.minimum
        reference_node = self
        parent_node = self.p
        while parent_node and reference_node is parent_node.right:
            parent_node = parent_node.p
            reference_node = parent_node
        return parent_node

    def predecessor(self):
        if self.left:
            return self.left.maximum
        reference_node = self
        parent_node = self.p
        while parent_node and reference_node is parent_node.left:
            parent_node = parent_node.p
            reference_node = parent_node
        return parent_node


class BinaryTree:
    """Binary tree"""

    def __init__(self, value=None):
        self.root = Node(value)

    def height(self):
        return self.root.height()

    def minimum(self):
        return self.root.minimum()

    def maximum(self):
        return self.root.maximum()

    def insert(self, value, node=None):
        # check for first insertion
        if not self.root.value:
            self.root.value = value
            return self.root

        # check for blank insertion
        node = self.root if not node else node

        if value < node.value:
            if not node.left:
                node.left = Node(value)
                node.left.p = node
                return node.left
            return self.insert(value, node.left)

        if not node.right:
            node.right = Node(value)
            node.right.p = node
            return node.right
        return self.insert(value, node.right)

    def search(self, value):
        pass

    def traverse_in_order(self):
        nodes = []
        def _traverse(node):
            if not node:
                return
            _traverse(node.left)
            nodes.append(node)
            _traverse(node.right)
        _traverse(self.root)
        return nodes


COLORS = {
    'RED': 1,
    'BLACK': 2,
}


class RedBlackTree(BinaryTree):
    """Red-Black tree"""

    def __init__(self):
        self.root = None

    def insert(self, value, node=None):
        self.root = self._insert(value, self.root)
        self.root.color = COLORS['RED']
        return self.root

    def _insert(self, value, node):
        insertion_root = node

        if node is None:
            new_node = Node(value)
            new_node.color = COLORS['RED']
            return new_node

        if value <= node.value:
            node.left = self._insert(value, node.left)
        else:
            node.right = self._insert(value, node.right)

        if node.right and node.right.color == COLORS['RED'] and \
           node.left and node.left.color == COLORS['BLACK']:
            insertion_root = self._left_rotate(node)

        if node.left and node.left.color == COLORS['RED'] and \
           node.left.left and node.left.left.color == COLORS['RED']:
            insertion_root = self._right_rotate(node)

        if node.right and node.right.color == COLORS['RED'] and \
           node.left and node.left.color == COLORS['RED']:
            node.right.color = COLORS['BLACK']
            node.left.color = COLORS['BLACK']

        return insertion_root

    def _left_rotate(self, node):
        """
            x
           / \
          a   y
             / \
            b   c
        """
        y = node.right
        if y:
            node.right = y.left
            y.left = node
            y.color = node.color
            node.color = COLORS['RED']
        return y

    def _right_rotate(self, node):
        """
            y
           / \
          x   c
         / \
        a   b
        """
        x = node.left
        if x:
            node.left = x.right
            x.right = node
            x.color = node.color
            node.color = COLORS['RED']
        return x

    def successor(self, node):
        pass

        # fixup:
        # right red, left black
        # insertion_root = lrotate

        # left red, left-left red
        # insertion_root = rrotate

        # left red, right red
        # flip colors of both


class Test(unittest.TestCase):
    TEST_LIST = [36, 11, 42, 24, 31, 12, 11, 32, 1, 27, 6, 39, 4, 12, 18, 13, 39, 21, 26, 1]

    def test_node_height(self):
        n = Node(5)
        n.left = Node(4)
        n.right = Node(6)
        n.right.right = Node(15)
        self.assertEqual(n.height(), 3)

    def test_binary_tree(self):
        t = BinaryTree()
        for r in self.TEST_LIST:
            t.insert(r)

        self.assertEqual(t.height(), 7)
        self.assertEqual(t.minimum(), 1)
        self.assertEqual(t.maximum(), 42)

    def test_red_black_tree(self):
        t = RedBlackTree()
        for r in self.TEST_LIST:
            t.insert(r)

        print(t.height())
        self.assertEqual(t.minimum(), 1)
        self.assertEqual(t.maximum(), 42)


unittest.main(verbosity=2)

# t = RedBlackTree()
# for r in Test.TEST_LIST:
#     t.insert(r)
# print([str(i) for i in t.traverse_in_order()])
# print(t.maximum())
